### PR TITLE
refactor: Split util/system into exception, shell, and fs-specific files

### PR DESCRIPTION
**Pull request by Empact**
This is an attempt to significantly improve code organization while making changes as simple as possible to review.

Here I split four files out of `util/system`:
* `util/exception`, holding `PrintExceptionContinue`
* `util/shell`, holding `ShellEscape`, `runCommand`
* `util/fs`, holding various file, folder, and path-specific functions (building on `fs.h` and `filesystem`)
* `util/args`, holding `ArgsManager` and related functions

The goal was to minimize review burden, so I made an effort to minimize the diffs and only modify the code when it was easy to confirm that the change would not have negative consequences.

That said, sourcing all of the windows-related headers was not straightforward, so I expect I'll need to touch this up to correct any issues there.

### CODE REVIEW

**Comment by Empact** [comment]
See prior discussion in #24455 /cc @dongcarl @ryanofsky

**Comment by Empact** [comment]
Thanks @dongcarl - I've got some issues to figure out, particularly on the `WIN32` side, and I'll be doing it via CI so it's going to mean some back and forth on this PR. E.g. the change is 5 commits but I'm going to have to play them out one by one to validate them.

File: src/util/shell.cpp [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+#pragma GCC diagnostic ignored "-Wnarrowing"
+#endif
+#include <boost/process.hpp>
+#include <memory>
+#include <ostream>
+#include <vector>
+#if defined(__GNUC__)
+#pragma GCC diagnostic pop
+#endif
+#endif // ENABLE_EXTERNAL_SIGNER
```
Comment (TheCharlatan) [inline src/util/shell.cpp:26]: AFAICT the build failures are coming from including these lines and I don't think they are needed, so I'm guessing this was included to debug a windows problem? I'm keen on moving this PR forward,. Is there some deeper problem here with running this patch on windows, or is it a compile-time issue?

File: src/util/exception.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#include <util/exception.h>
+
+#include <logging.h>
+#include <tinyformat.h>
+
+#include <iostream>
+#include <typeinfo>
+
+#ifdef WIN32
+#include <compat/compat.h>
```
Comment (fanquake) [inline src/util/exception.cpp:15]: In f12cd1d77bc062f1313f98a8aaea8afd96c4f52e:
Why is compat being included here? if it's for a `windows.h` include (which isn't in compat anyways), then I'd rather we just include `windows.h`.

File: src/util/shell.cpp (hunk 1 above)
Comment (fanquake) [reply src/util/shell.cpp:26]: in e5243d3676136e4b0a590e76a07c86e62ca7d1f7:
>  and I don't think they are needed,

Yea, there is no Boost usage in this file (and is no-longer any in the util library), so the entire `ENABLE_EXTERNAL_SIGNER` chunk here should be removed.

File: src/util/shell.cpp [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+#include <tinyformat.h>
+#include <univalue.h>
+#include <util/string.h>
+
+#include <cstdlib>
+#include <stdexcept>
+#include <string>
+
+#ifdef WIN32
+#include <compat/compat.h>
```
Comment (fanquake) [inline src/util/shell.cpp:38]: In e5243d3676136e4b0a590e76a07c86e62ca7d1f7:
If this is for `windows.h`, would rather just include that.

File: src/util/shell.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+#ifndef BITCOIN_UTIL_SHELL_H
+#define BITCOIN_UTIL_SHELL_H
+
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
+#include <string>
+
+class UniValue;
```
Comment (fanquake) [inline src/util/shell.h:15]: In e5243d3676136e4b0a590e76a07c86e62ca7d1f7: Why is `UniValue` in here?

File: test/lint/lint-circular-dependencies.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ def main():
 
     os.chdir(CODE_DIR)
     files = subprocess.check_output(
-        ['git', 'ls-files', '--', '*.h', '*.cpp'],
+        ["git", "ls-files", "--", "*.h", "*.cpp"],
```
Comment (fanquake) [inline test/lint/lint-circular-dependencies.py:40]: In 9ff4bb56baa509c587698cb3a2877dfd4ff6d136:
This seems unrelated to this commit, and is not mentioned in the commit message.

**Review by christ79ma — APPROVED** [review]

**Comment by fanquake** [comment]
> I'm still keen on moving this work forward. If you don't have the time right now, I'll attribute the commits to you and try to get my branch merged instead.

Yep, it would be good to start getting some of this merged.

**Comment by fanquake** [comment]
Going to close this for now, given that it's being split out, and there's a [thumbs up from the author above](https://github.com/bitcoin/bitcoin/pull/25152#issuecomment-1458326647).
